### PR TITLE
Check if form calls IsSubmitted

### LINKED ISSUE #9826: Mautic 4: Unable to merge companies
⚠️ This relates to Mautic 4 Alpha

## Bug Description
I am unable to merge companies - the merge happened but it threw an error.

NB the first time it happened I ended up with a contact of the same name as the company I was merging.

Edit: To rule out sample data being the issue I tried to merge two companies that I created manually and the error still occurred but the random contact was **not** created. So that part seems to be an issue with the fixtures for contacts/companies maybe related to #9090.

| Q   | A
| --- | ---
| Mautic version | 4.0-Alpha1
| PHP version | 7.4
| Browser | Chrome

### Steps to reproduce
1. Install Mautic 4.0-alpha1 and install sample data
2. Merge the company Amazon into Apple
3. Notice the error - the merge has happened and the companies have been merged / contacts updated.

https://youtu.be/qXYFLaLQ_RM screen capture.
 
### Log errors

```
[2021-03-25 14:58:02] mautic.DEBUG: COMPANY: Merging companies  
[2021-03-25 14:58:04] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Form\Exception\LogicException: "Cannot check if an unsubmitted form is valid. Call Form::isSubmitted() before Form::isValid()." at /var/www/html/vendor/symfony/form/Form.php line 745 {"exception":"[object] (Symfony\\Component\\Form\\Exception\\LogicException(code: 0): Cannot check if an unsubmitted form is valid. Call Form::isSubmitted() before Form::isValid(). at /var/www/html/vendor/symfony/form/Form.php:745)
[stacktrace]
#0 /var/www/html/app/bundles/CoreBundle/Controller/AbstractFormController.php(169): Symfony\\Component\\Form\\Form->isValid()
#1 /var/www/html/app/bundles/LeadBundle/Controller/CompanyController.php(405): Mautic\\CoreBundle\\Controller\\AbstractFormController->isFormValid(Object(Symfony\\Component\\Form\\Form))
#2 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(158): Mautic\\LeadBundle\\Controller\\CompanyController->editAction(3, false)
#3 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(80): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 2)
#4 /var/www/html/vendor/symfony/framework-bundle/Controller/ControllerTrait.php(96): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 2)
#5 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(370): Symfony\\Bundle\\FrameworkBundle\\Controller\\Controller->forward('MauticLeadBundl...', Array, Array)
#6 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(300): Mautic\\CoreBundle\\Controller\\CommonController->ajaxAction(Array)
#7 /var/www/html/app/bundles/LeadBundle/Controller/CompanyController.php(936): Mautic\\CoreBundle\\Controller\\CommonController->postActionRedirect(Array)
#8 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(458): Mautic\\LeadBundle\\Controller\\CompanyController->mergeAction('2', '')
#9 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(158): Mautic\\CoreBundle\\Controller\\CommonController->executeAction('merge', '2', 0, '')
#10 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(80): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#11 /var/www/html/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#12 /var/www/html/app/AppKernel.php(104): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#13 /var/www/html/app/middlewares/CORSMiddleware.php(91): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#14 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(43): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#15 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(61): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#16 /var/www/html/app/middlewares/VersionCheckMiddleware.php(67): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#17 /var/www/html/app/middlewares/TrustMiddleware.php(51): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#18 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#19 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#20 /var/www/html/index_dev.php(25): Stack\
un(Object(Stack\\StackedHttpKernel))
#21 {main}
"} 
```

### PR DESCRIPTION
**Pull request by mabumusa1**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" 
| Bug fix?                               | yes
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes/no
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #9826

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2.

### CODE REVIEW

**Review by RCheesley — DISMISSED** [review]
Confirmed the bug before applying the PR - 500 error. After applying the PR I was able to merge the two companies without an error and the contacts are correctly assigned to the company as a result. Thanks @mabumusa1 ! 🚀 🎉

File: app/bundles/CoreBundle/Controller/AbstractFormController.php [hunk 1]
```diff
@@ -166,7 +166,9 @@ protected function isFormValid(Form $form, array $data = null)
         //bind request to the form
         $form->handleRequest($this->request);
 
-        return $form->isValid();
+        if ($form->isSubmitted()) {
+            return $form->isValid();
+        }
```
Comment (escopecz) [inline app/bundles/CoreBundle/Controller/AbstractFormController.php]: The method should return bool. With the suggested change there is a chance the method won't return anything. I'd suggest this:
```suggestion
        return $form->isSubmitted() && $form->isValid();
```

**Review by escopecz — CHANGES_REQUESTED** [review]

File: app/bundles/CoreBundle/Controller/AbstractFormController.php (hunk 1 above)
Comment (mabumusa1) [reply app/bundles/CoreBundle/Controller/AbstractFormController.php]: @escopecz I agree with you, In the first comment I made the function return null, then I removed it. I think your solution is better to make sure we always return the right value.

Thanks

**Review by escopecz — APPROVED** [review]
Thanks! 👍

**Comment by RCheesley** [comment]
@all-contributors add mabumusa1 for code

**Comment by RCheesley** [comment]
@all-contributors add mabumusa1 for code

**Comment by RCheesley** [comment]
@all-contributors add @mabumusa1 for code

**Comment by RCheesley** [comment]
@all-contributors add @mabumusa1 for code

**Comment by RCheesley** [comment]
@all-contributors add @mabumusa1 for code

**Pull request merged**